### ISSUE TITLE
Strategy for older Python and dependencies support

### ISSUE THREAD

**Issue by martinfleis** [geopandas/geopandas#1457]
We are now quite generous in what we support, having still env with Python 3.5, pandas 0.23 and numpy 1.12 on our CI. We should have a clear strategy when to drop which.

Numpy has a nice [recommendation](https://numpy.org/neps/nep-0029-deprecation_policy.html) of the minimal version which should be supported. For numpy itself it is now 1.15 (current is 1.18). 

For Python they recommend 42 months of support since the release of each minor version, which would mean supporting only Python 3.7+ from June 23. 

For numpy, it the support window is set to 24 months, which would be numpy 1.15 now.

I haven't found similar recommendation for pandas, but I feel that using the same 24 months makes sense, which means that support for pandas 0.23 just expired.

After 0.8.0 release, I would drop support for Python 3.5, and support:

- Python 3.6+
- pandas 0.24+ 
- numpy 1.15+
- shapely 1.6+ (change from 1.5, which was released in 2014)
- fiona 1.8+ (change from 1.7, released in 2016)
- matplotlib 2.2.0 (change from 2.0, released in 2017)

We can announce it (at least Python) in the release notes of 0.8. There is no point in dropping support for 0.8, since the whole code already does support all (old) above. But after the release we could lower the maintenance cost a bit.

**Comment by jorisvandenbossche** [comment]
Yes, I agree we should keep our supported versions a bit more "up to date" (we eg could have dropped Python 3.5 support a while ago). The numpy NEP is a good reference, and also pandas mostly follows that.  

Your proposed list of versions looks good to me, as well as announcing 0.8 as the last version to support the older versions, and dropping python 3.5 support after 0.8.

**Comment by snowman2** [comment]
EOL is December 2021: https://devguide.python.org/#status-of-python-branches

- numpy dropped support June 2020 [ref](https://numpy.org/neps/nep-0029-deprecation_policy.html)
- pandas dropped support June 2020 [ref](https://github.com/pandas-dev/pandas/issues/34472)
- xarray dropped support in 0.17.0 (http://xarray.pydata.org/en/stable/whats-new.html#breaking-changes)
- conda-forge support will be dropped soon [ref](https://github.com/conda-forge/rasterio-feedstock/pull/184#issuecomment-767161877)

**Comment by snowman2** [comment]
Pyproj issue: https://github.com/pyproj4/pyproj/issues/790

**Comment by martinfleis** [comment]
Thanks for the info. I'll make a PR deprecating support of Python 3.6 and older versions of our dependencies during the weekend (I was planning to anyway).

Will follow NEP assuming that 0.10.0 will come out in July (hence calculating support from July 21, not today).
